src/mock/quick_p2p/tests.rs

```diff
@@ -527,6 +523,22 @@ impl Agent {
     fn expect_none(&self) {
         assert_eq!(self.rx.try_recv(), Err(TryRecvError::Empty));
     }
+
+    fn received_messages(&self, src_addr: &SocketAddr) -> BTreeSet<Bytes> {
+        let mut received_messages = BTreeSet::new();
+        while let Ok(Event::NewMessage { peer_addr, msg }) = self.rx.try_recv() {
+            assert_eq!(peer_addr, *src_addr);
+            let _ = received_messages.insert(msg);
+        }
+        received_messages
+    }
+}
+
+fn expected_messages_received(expected: Vec<Bytes>, mut received: BTreeSet<Bytes>) {
```

Comment:
Why no take `expected` by `BTreeSet` as well. Then you would be able to do just `assert_eq!(expected, received)`.

Reply:
The `expected` was a Vec in the order of sending out messages.
We need to convert it into BTreeSet somewhere anyway.
I made it here as it can also diff the expected and received explicitly by type.

Reply:
There are two issue with this (both relatively minor though):
1. The parameter `expected: Vec<Bytes>` suggest that the order is also what is expected. If the parameter was `expected: BTreeSet<Bytes>` then it would be more clear that we only care about the elements being the same, but not their order.
2. The implementation of this function could be changed to just `assert_eq!(expected, received)` which is not only shorter, but it would produce more useful error messages in case of failure.

Reply:
but we still need to convert that send_out msgs (which is a `vec` into `BTreeSet`) somewhere, and with more tests, such convertion may happen multiple times.
How about:
```
fn expected_messages_received(sent: Vec<Bytes>, mut received: BTreeSet<Bytes>) {
    let expected: BTreeSet<Bytes> = sent.iter().cloned().collect();
    assert_eq!(expected, received);
}
```
so basicaly the convertion got capsulated in this function?

Reply:
That works. You don't need the `mut` though.